packages/keystone/adapterCache.js

```diff
@@ -0,0 +1,335 @@
+        if (!this.debugMode) { return }
+        if (!this.cacheHistory[table]) {
+            this.cacheHistory[table] = []
+        }
+        const copiedCache = cloneDeep(cache)
+        this.cacheHistory[table].push({
+            cache: copiedCache,
+            cacheByTable: copiedCache[table],
+            type: event.type,
+            event: event,
+            dateTime: new Date().toLocaleString(),
+            number: this.totalRequests,
+        })
+        this.cacheCallHistory.push({
+            name: table,
+            eventType: event.type,
+            dateTime: new Date().toLocaleString(),
+            number: cloneDeep(this.totalRequests),
+        })
+        this.cacheHistory.lastTableUpdated = table
+    }
+
+    getCacheEvent ({ type, key, table, result }) {
+        return ({
+            type: type,
+            table: table,
+            string: `
+            🔴 STAT: ${this.cacheHits}/${this.totalRequests}\r\n
+            🔴 RKEY: ${key}\r\n
+            🔴 TYPE: ${type}\r\n
+            🔴 RESP: ${result}\r\n
+            `,
+        })
+    }
+
+    logEvent ({ event }) {
+        if (!this.logging) { return }
+        logger.info(event.string)
+    }
+
+    async prepareMiddleware ({ keystone, dev, distDir }) {
+        if (this.enabled) {
+            await patchKeystoneAdapterWithCacheMiddleware(keystone, this)
+            logger.info('ADAPTER_CACHE: Adapter level cache ENABLED')
+        } else {
+            logger.info('ADAPTER_CACHE: Adapter level cache NOT ENABLED')
+        }
+    }
+}
+
+/**
+ * Patches an internal keystone adapter adding cache functionality
+ * @param keystone
+ * @param {AdapterCacheMiddleware} middleware
+ * @returns {Promise<void>}
+ */
+async function patchKeystoneAdapterWithCacheMiddleware (keystone, middleware) {
+    const keystoneAdapter = keystone.adapter
+
+    const cache = middleware.cache
+    const includeAllLists = middleware.includeAllLists
+    const excludedLists = middleware.excludedLists
+    const includedLists = middleware.includedLists
+
+    const listAdapters = Object.values(keystoneAdapter.listAdapters)
+
+    for (const listAdapter of listAdapters) {
+
+        const listName = listAdapter.key
+        cache[listName] = {}
+
+        const fields = listAdapter.fieldAdaptersByPath
+        if (!fields[UPDATED_AT_FIELD] || !fields[UPDATED_AT_FIELD]) {
+            logger.info(`ADAPTER_CACHE: Cache is NOT enabled for list: ${listName} -> No ${UPDATED_AT_FIELD} field`)
+        }
+
+        if (excludedLists.includes(listName) ||
+            !includeAllLists && !includedLists.includes(listName)) {
+            logger.info(`ADAPTER_CACHE: Cache is NOT enabled for list: ${listName} -> Cache is not included by config`)
+            continue
+        }
+
+        logger.info(`ADAPTER_CACHE: Cache is enabled for list: ${listName}`)
+
+        const originalItemsQuery = listAdapter._itemsQuery
+        listAdapter._itemsQuery = async ( args, opts ) => {
```

Comment:
it's better to use public methods

Reply:
Yep, now using public methods substitution